library/ssl_msg.c

```diff
@@ -536,76 +536,39 @@ static void ssl_mac( mbedtls_md_context_t *md_ctx,
 }
 #endif /* MBEDTLS_SSL_PROTO_SSL3 */
 
-#define SSL_RECORD_AEAD_NONCE_UNKNOWN 0u
-#define SSL_RECORD_AEAD_NONCE_CONCAT  1u
-#define SSL_RECORD_AEAD_NONCE_XOR     2u
-
-static int ssl_transform_get_nonce_mode( mbedtls_ssl_transform const *transform )
+static int ssl_transform_aead_dynamic_iv_is_explicit(
+                                mbedtls_ssl_transform const *transform )
 {
-#if defined(MBEDTLS_CHACHAPOLY_C)
-    if( transform->ivlen == 12 && transform->fixed_ivlen == 12 )
-    {
-        return( SSL_RECORD_AEAD_NONCE_XOR );
-    }
-#endif /* MBEDTLS_CHACHAPOLY_C */
-
-#if defined(MBEDTLS_GCM_C) || defined(MBEDTLS_CCM_C)
-    if( transform->ivlen == 12 && transform->fixed_ivlen == 4 )
-    {
-        return( SSL_RECORD_AEAD_NONCE_CONCAT );
-    }
-#endif /* MBEDTLS_GCM_C || MBEDTLS_CCM_C */
-
-    return( SSL_RECORD_AEAD_NONCE_UNKNOWN );
+    return( transform->ivlen != transform->fixed_ivlen );
 }
 
-/* Preconditions:
- * - If mode == SSL_RECORD_AEAD_NONCE_CONCAT, then
- *     dst_nonce_len == fixed_iv_len + dynamic_iv_len
- * - If mode == SSL_RECORD_AEAD_NONCE_XOR, then
- *     dst_nonce_len == fixed_iv_len &&
- *     dynamic_iv_len < dst_nonce
+
+/* Compute IV := ( fixed_iv || 0 ) XOR ( 0 || dynamic_IV )
```

Comment:
This implies the constraints on the lengths of the parameters, but it still might worth noting that the function assumes that `dst_iv_len >= fixed_iv_len` and `dst_iv_len >= dynamic_iv_len`. These are needed for safe operation.

Reply:
I've documented the precondition.